### PR TITLE
Pass device in Logits Processor's init

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

This PR adds the ability to pass in device when initializing `LogitsProcessors` and is one more step towards `compile` compatibility. 

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante

### CODE REVIEW

**Review by gante — COMMENTED** [review]
Overall notes before going to details:
1. In the processors that take `eos_token_id` as input: see https://github.com/huggingface/transformers/pull/29788. In this PR, the special tokens are treated as tensors by default, solving most of the needed changes. I would rebase this PR on `main` after that PR is merged, as some of the changes here will become redundant :)
2. On the processors that don't need to use `device`, such as `TemperatureLogitsWarper` -- let's not add unused arguments. Clean interfaces are important 🧼 (unless there are significant benefits from standardizing them)
3. Let's not throw a warning when the device is not passed and tensors are initialized on CPU. A `.to` operation is not that expensive :)

**Comment by zucchini-nlp** [comment]
1. Cool, I did not notice that
2 and 3. Okay, thought we need it for consistency like we had with other new args in public classes. Will remove it and rebase later

**Comment by zucchini-nlp** [comment]
Not stale

**Comment by zucchini-nlp** [comment]
This PR now can be reviewed. Rebased main and updated the changes. All the tests from `RUN_SLOW=1 pytest tests/generation` are passing on my end

**Review by gante — APPROVED** [review]
LGTM, thank you for improving `generate` :D

File: src/transformers/models/whisper/generation_whisper.py
Comment (zucchini-nlp) [reply src/transformers/models/whisper/generation_whisper.py:549]: Yeah, for some reason I thought `input_features` can be `None` as it's optional in the docstring 😅

Comment (gante) [reply src/transformers/models/whisper/generation_whisper.py]: @zucchini-nlp Good point. Whisper needs `input_features` OR some encoder inputs. We should encode this path too. (It's still preferable to `next(self.parameters()).device`, as we want to infer the device from the inputs)

**Comment by zucchini-nlp** [comment]
@gante Ah I forgot whisper is encoder-decoder. Oke, now it infers device from one of the inputs passed by the user. 

Tagging the core maintainer for review, failing tests are not related and seems given `@unittest.skip("tfp is not defined even if installed. FIXME @Arthur in a followup PR!")`  :)

**Comment by ArthurZucker** [comment]
How could the bot come 🤣 anyways on it!

File: src/transformers/generation/logits_process.py
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:148]: I'd argue that we can just set it to "cpu" in the arg no?

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1740]: why is this required ?

File: src/transformers/generation/utils.py
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1]: should be use self.device? or lm_head.device? (which is not always there but still)

File: tests/generation/test_logits_process.py
Comment (ArthurZucker) [inline tests/generation/test_logits_process.py:1]: I think we need to make sure dive placement on multi GPU works, might already be tested  !

**Review by ArthurZucker — COMMENTED** [review]
Overall LGTM, not sure input_ids device is the always the best, and we need a small test to see which feature is enable by this potentially!

File: src/transformers/generation/logits_process.py
Comment (zucchini-nlp) [reply src/transformers/generation/logits_process.py:148]: This is mostly for users who use/pass LogitsProcessor as a standalone kwarg, because 'generate()' takes care that device is not None.

I think we should raise warning for BC saying users to pass-in the device, but let's ask @gante if he's okay with it. If I am not misunderstanding, we shouldn't raise warnings 🤔 

> Let's not throw a warning when the device is not passed and tensors are initialized on CPU. A .to operation is not that expensive :)

File: src/transformers/generation/utils.py
Comment (zucchini-nlp) [reply src/transformers/generation/utils.py:1740]: Right! I thought that it was me who changed to `inputs_tensor` and was trying to revert 😆 I'll revert it back, no difference whichever tensor we use here

File: src/transformers/generation/logits_process.py
Comment (ArthurZucker) [reply src/transformers/generation/logits_process.py:148]: yeah, don't think it's a problem to silently do this

File: tests/generation/test_logits_process.py
Comment (ArthurZucker) [reply tests/generation/test_logits_process.py:1]: got it. Any how LGTM

File: src/transformers/generation/logits_process.py
Comment (ArthurZucker) [reply src/transformers/generation/logits_process.py:148]: Down to just default to CPU which was already the behaviour by default before this PR no?

Comment (zucchini-nlp) [reply src/transformers/generation/logits_process.py:148]: ahh my bad, didn't read carefully the first comment. Setting in the arg as default is better, right 

My concern is that before this PR, we were placing these on `scores.device` during "_ [call]_ " , but anyway I still get lost at when to do BC deprecation and when to not do 😄 https://github.com/huggingface/transformers/blob/03935d300d60110bb86edb49d2315089cfb19789/src/transformers/generation/logits_process.py#L155

**Review by ArthurZucker — APPROVED** [review]
could you rebase your branch ? (format changes seems unrelated?)

**Comment by zucchini-nlp** [comment]
Oke, rebased main and the unnecessary formatting is removed. Will merge as I guess we don't need to add warnings :)

**Pull request merged**
